Source/Core/Core/WiiUtils.cpp

```diff
@@ -74,4 +97,392 @@ bool InstallWAD(const std::string& wad_path)
   DiscIO::NANDContentManager::Access().ClearCache();
   return true;
 }
+
+class SystemUpdater final
+{
+public:
+  UpdateResult DoOnlineUpdate(UpdateCallback update_callback);
+
+private:
+  struct TitleInfo
+  {
+    u64 id;
+    u16 version;
+  };
+
+  struct Response
+  {
+    std::string content_prefix_url;
+    std::vector<TitleInfo> titles;
+  };
+
+  std::string GetDeviceRegion();
+  std::string GetDeviceId();
+
+  Response GetSystemTitles();
+  Response ParseTitlesResponse(const std::vector<u8>& response) const;
+
+  bool InstallTitleFromNUS(const std::string& prefix_url, const TitleInfo& title,
+                           std::unordered_set<u64>* updated_titles);
+
+  // Helper functions to download contents from NUS.
+  std::pair<IOS::ES::TMDReader, std::vector<u8>> DownloadTMD(const std::string& prefix_url,
+                                                             const TitleInfo& title);
+  std::pair<std::vector<u8>, std::vector<u8>> DownloadTicket(const std::string& prefix_url,
+                                                             const TitleInfo& title);
+  std::optional<std::vector<u8>> DownloadContent(const std::string& prefix_url,
+                                                 const TitleInfo& title, u32 cid);
+
+  UpdateCallback m_update_callback;
+  IOS::HLE::Kernel m_ios;
+  Common::HttpRequest m_http{0};
+};
+
+std::string SystemUpdater::GetDeviceRegion()
+{
+  // Try to determine the region from an installed system menu.
+  const auto tmd = m_ios.GetES()->FindInstalledTMD(TITLEID_SYSMENU);
+  if (tmd.IsValid())
+  {
+    const DiscIO::Region region = tmd.GetRegion();
+    static const std::map<DiscIO::Region, std::string> regions = {
+        {DiscIO::Region::NTSC_J, "JPN"},
+        {DiscIO::Region::NTSC_U, "USA"},
+        {DiscIO::Region::PAL, "EUR"},
+        {DiscIO::Region::NTSC_K, "KOR"},
+        {DiscIO::Region::UNKNOWN_REGION, "EUR"}};
+    return regions.at(region);
+  }
+
+  // Try to determine the region from the setting.txt file.
+  // We check that file after the system menu, because Dolphin likes to change the region
+  // automatically without warning the user when launching a disc game from the game list.
+  SettingsHandler gen;
+  const std::string settings_file_path =
+      Common::GetTitleDataPath(TITLEID_SYSMENU, Common::FROM_CONFIGURED_ROOT) + WII_SETTING;
+  if (File::Exists(settings_file_path) && gen.Open(settings_file_path))
+  {
+    const std::string region = gen.GetValue("AREA");
+    if (!region.empty())
+      return region;
+  }
+
+  // Fall back to EUR.
+  return "EUR";
+}
+
+std::string SystemUpdater::GetDeviceId()
+{
+  u32 ios_device_id;
+  m_ios.GetES()->GetDeviceId(&ios_device_id);
+  return StringFromFormat("%" PRIu64, (u64(1) << 32) | ios_device_id);
+}
+
+constexpr const char* GET_SYSTEM_TITLES_REQUEST_PAYLOAD = R"END(\
```

Comment:
Do you intentionally want this XML file to start with "backslash newline"? Usually the XML-declaration should (or rather, must) be the first thing to appear, sans a BOM if any.